### ISSUE TITLE
Type checking breaks under when importing a `Union[str, TypedDict]`

### ISSUE THREAD

**Issue by multimeric** [microsoft/pyright#3796]
**Describe the bug**
When you define a type union which is `Union[str, T]` where `T` is a `TypedDict` subclass, *and* you import the type union from another module, then pyright fails to reconcile a dict literal with the type union. If any of these conditions are removed, it seems to work.

**To Reproduce**

Here I define `query.py`:
```python
from typing import Iterable, Union
from typing_extensions import NotRequired, TypeAlias, TypedDict


class QueryDict(TypedDict):
    and_: NotRequired[Iterable["Query"]]
    field: NotRequired[str]


Query: TypeAlias = Union[QueryDict, str]
```
Then, in the same directory, `import.py`:
```python
from query import Query

x: Query = {"and_": ["", {"field": "foo"}]}
y: Query = {"and_": [{"field": "bar"}, {"field": "foo"}]}
z: Query = {"and_": ["", ""]}
```

Then `pyright import.py`, gives:
```
/home/migwell/Programming/PylanceBug/import.py
  /home/migwell/Programming/PylanceBug/import.py:3:12 - error: Expression of type "dict[str, list[str | dict[str, str]]]" cannot be assigned to declared type "Query"
    Type "dict[str, list[str | dict[str, str]]]" cannot be assigned to type "Query"
      "dict[str, list[str | dict[str, str]]]" is incompatible with "QueryDict"
      "dict[str, list[str | dict[str, str]]]" is incompatible with "str" (reportGeneralTypeIssues)
  /home/migwell/Programming/PylanceBug/import.py:4:12 - error: Expression of type "dict[str, list[dict[str, str]]]" cannot be assigned to declared type "Query"
    Type "dict[str, list[dict[str, str]]]" cannot be assigned to type "Query"
      "dict[str, list[dict[str, str]]]" is incompatible with "QueryDict"
      "dict[str, list[dict[str, str]]]" is incompatible with "str" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations
```

**Expected behavior**
There should be no type error detected in the code below.

**Additional context**
Notably, if I put the exact same definition of `x`, `y`, `z` into `query.py`, `pyright` finds no errors. Also, note that the definition of `z` works fine, so the string type is being understood correctly.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

The problem was related to the way pyright handles circular references for type aliases. `QueryDict` referred to the type alias `Query`, which refers to `QueryDict`.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.266, which I just published. It will also be included in a future release of pylance.

**Comment by multimeric** [comment]
Hmm, the latest pylance seems to have broken the type inference for this scenario which is evident in the example I've used above. Is that anything to do with your changes, or should I report this in the pylance repo? For what it's worth, `pylance` from the CLI works correctly.
```
Expression of type "dict[str, list[str | dict[str, str]]]" cannot be assigned to declared type "Query"
  Type "dict[str, list[str | dict[str, str]]]" cannot be assigned to type "Query"
    "dict[str, list[str | dict[str, str]]]" is incompatible with "QueryDict"
    "dict[str, list[str | dict[str, str]]]" is incompatible with "str"Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md#diagnostic-severity-rules)
```

**Comment by erictraut** [comment]
The latest production release of pylance does not yet include the changes from pyright 1.1.266. If you switch to the prerelease (insiders) build of pylance, you should see this issue fixed.
